File: apps/worker/cairn_worker/llm.py
"""call_model(): the one wrapper every Python model call goes through.

Writes the same llm_calls rows as packages/intel/llm.ts. Keep PRICING in sync
with the TS file; the Token Company table is built from this ledger.

OpenAI Responses API. Cached input is automatic on a stable prefix; pass
``prompt_cache_key`` per chapter so the 20-node batches route to one cache.
"""

from __future__ import annotations

import json
import math
import os
import time
from dataclasses import dataclass
from typing import Any
from uuid import UUID

import psycopg
from openai import OpenAI
from openai.types.responses import Response, ResponseInputParam

from cairn_worker.config import settings

# USD per million tokens, from the OpenAI pricing page on 2026-09-19. Sol is on
# promotional pricing through at least 2026-11-21; re-pull at the venue. Cache
# writes bill at 1.25x input but the SDK's usage object does not report them.
PRICING: dict[str, dict[str, float]] = {
    "gpt-5.6-luna": {"input": 0.2, "cached_input": 0.02, "cache_write": 0.25, "output": 1.2},
    "gpt-5.6-sol": {"input": 4.0, "cached_input": 0.4, "cache_write": 5.0, "output": 20.0},
    "text-embedding-3-small": {"input": 0.02, "cached_input": 0.02, "cache_write": 0.0, "output": 0.0},
}

_client: OpenAI | None = None


def _openai() -> OpenAI:
    global _client
    if _client is None:
        _client = OpenAI(max_retries=0)
    return _client


@dataclass(frozen=True)
class LedgerUsage:
    """Ledger token columns. input_tokens EXCLUDES cache reads so the columns are disjoint."""

    input_tokens: int
    output_tokens: int
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0


def to_ledger_usage(response: Response) -> LedgerUsage:
    """OpenAI's input_tokens includes cached_tokens; split them for the ledger."""
    u = response.usage
    if u is None:
        raise ValueError("Missing usage; billing reconciliation required")
    cached = u.input_tokens_details.cached_tokens if u.input_tokens_details else 0
    writes = getattr(u.input_tokens_details, "cache_write_tokens", 0) or 0
    values = (u.input_tokens - cached - writes, u.output_tokens, cached, writes)
    if any(not isinstance(n, int) or n < 0 for n in values):
        raise ValueError("Invalid token usage")
    return LedgerUsage(*values)


def estimate_cost_usd(model: str, u: LedgerUsage) -> float:
    p = PRICING.get(model)
    if not p:
        raise ValueError(f"No verified pricing for model: {model}")
    if u.input_tokens + u.cache_read_tokens + u.cache_write_tokens > 272000:
        raise ValueError("Long-context pricing requires reconciliation")
    return (
        u.input_tokens * p["input"]
        + u.cache_read_tokens * p["cached_input"]
        + u.cache_write_tokens * p["cache_write"]
        + u.output_tokens * p["output"]
    ) / 1_000_000


def call_model(
    conn: psycopg.Connection | None,
    *,
    stage: str,
    input: str | ResponseInputParam,
    instructions: str | None = None,
    model: str | None = None,
    prompt_cache_key: str | None = None,
    json_schema: dict[str, Any] | None = None,
    max_output_tokens: int = 8192,
    doc_id: UUID | None = None,
    corpus_id: UUID | None = None,
    meta: dict[str, Any] | None = None,
) -> tuple[str, Response]:
    """Returns (output_text, full response). Logs one ledger row.

    Keep the chapter context in ``instructions`` byte-identical across a batch
    and pass one ``prompt_cache_key`` per chapter so cached input hits (A2, C5).
    ``json_schema`` = {"name": ..., "schema": {...}} enables strict JSON output.
    """
    model = model or settings.model_fast
    if os.environ.get("CAIRN_LIVE_API") != "1":
        raise RuntimeError("Live API calls are disabled")
    if conn is None or not conn.autocommit:
        raise RuntimeError("Paid calls require an autocommit ledger connection")
    if not os.environ.get("OPENAI_API_KEY"):
        raise RuntimeError("OPENAI_API_KEY is not configured in the process environment")
    estimate_cost_usd(model, LedgerUsage(0, 0))
    if not isinstance(max_output_tokens, int) or not 0 < max_output_tokens <= 128000:
        raise ValueError("Invalid maximum output tokens")
    if not isinstance(input, str):
        if not all(isinstance(item, dict) and isinstance(item.get("content"), str) and "role" in item for item in input):
            raise ValueError("Only text model inputs are supported")
    if len(json.dumps([input, instructions, json_schema], ensure_ascii=False).encode()) > 240000:
        raise ValueError("Input exceeds short-context budget")
    price = PRICING[model]
    maximum = math.ceil(272000 * max(price["input"], price["cache_write"]) + max_output_tokens * price["output"])
    reservation_row = conn.execute("SELECT cairn_reserve(%s,%s,%s) AS id", (maximum, stage, model)).fetchone()
    reservation = reservation_row["id"] if isinstance(reservation_row, dict) else reservation_row[0]
    kwargs: dict[str, Any] = {}
    if instructions is not None:
        kwargs["instructions"] = instructions
    if prompt_cache_key is not None:
        kwargs["prompt_cache_key"] = prompt_cache_key
    if json_schema is not None:
        kwargs["text"] = {"format": {"type": "json_schema", "strict": True, **json_schema}}

    started = time.perf_counter()
    response = _openai().responses.create(model=model, input=input, max_output_tokens=max_output_tokens, service_tier="default", **kwargs)
    latency_ms = int((time.perf_counter() - started) * 1000)

    usage = to_ledger_usage(response)
    if conn is not None:
        log_call(conn, stage=stage, model=model, usage=usage, latency_ms=latency_ms,
                 cost_usd=estimate_cost_usd(model, usage), doc_id=doc_id, corpus_id=corpus_id,
                 meta={**(meta or {}), "reservation_id": str(reservation), "response_id": response.id})
    # Cache-write usage is not exposed by every SDK: settle a conservative upper bound.
    actual = math.ceil((usage.input_tokens + usage.cache_write_tokens) * max(price["input"], price["cache_write"])
                       + usage.cache_read_tokens * price["cached_input"] + usage.output_tokens * price["output"])
    settled = conn.execute("SELECT cairn_settle(%s,%s) AS ok", (reservation, actual)).fetchone()
    if not (settled["ok"] if isinstance(settled, dict) else settled[0]):
        raise RuntimeError("Cost exceeded reservation; budget blocked")
    if response.status != "completed" or not response.output_text.strip():
        raise RuntimeError("Model response incomplete, refused, or empty")
    if json_schema is not None:
        json.loads(response.output_text)
    return response.output_text, response


def log_call(
    conn: psycopg.Connection,
    *,
    stage: str,
    model: str,
    usage: LedgerUsage,
    latency_ms: int,
    cost_usd: float,
    doc_id: UUID | None = None,
    corpus_id: UUID | None = None,
    meta: dict[str, Any] | None = None,
) -> None:
    conn.execute(
        """
        INSERT INTO llm_calls
          (stage, model, input_tokens, output_tokens, cache_read_tokens, cache_write_tokens,
           latency_ms, cost_usd, corpus_id, doc_id, meta)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s::jsonb)
        """,
        (
            stage, model, usage.input_tokens, usage.output_tokens,
            usage.cache_read_tokens, usage.cache_write_tokens,
            latency_ms, cost_usd, corpus_id, doc_id, json.dumps(meta or {}),
        ),
    )

File: packages/intel/llm.ts
/**
 * callModel(): the one wrapper every TypeScript model call goes through.
 * Every call writes one row to `llm_calls` (see packages/contracts/schema.sql).
 * The Python worker's llm.py writes the same rows for its own calls.
 *
 * OpenAI Responses API. Cached input is automatic on a stable prefix; pass a
 * `promptCacheKey` per chapter so batched calls route to the same cache (C5).
 *
 * C0 owns this file. Phase C0 acceptance: `pnpm --filter @quod/intel cost-report`
 * prints totals by stage, and a synthetic run of 50 calls is logged accurately.
 */
import OpenAI from "openai";
import { createHash } from "node:crypto";
import { db } from "@quod/contracts/db";
import { requireLiveBudget, reserveApiSpend, settleApiSpend } from "./budget";

/** Stage tag on every ledger row. Extend freely; keep names stable once used. */
export type Stage =
  | "segment"
  | "edges"
  | "resolve"
  | "instantiate"
  | "trace"
  | "embed"
  | "search"
  | "voice"
  | "eval";

/** Model roles from the spec. Override with MODEL_FAST / MODEL_QUALITY / EMBEDDING_MODEL. */
export const MODELS = {
  /** Bulk classification: segmentation, edge extraction. */
  fast: process.env.MODEL_FAST ?? "gpt-5.6-luna",
  /** The two calls a judge sees: instantiation and resolution adjudication. */
  quality: process.env.MODEL_QUALITY ?? "gpt-5.6-sol",
  /** Dense vectors for the Elasticsearch hybrid index (C2). */
  embedding: process.env.EMBEDDING_MODEL ?? "text-embedding-3-small",
} as const;

/**
 * USD per million tokens, checked against https://developers.openai.com/api/docs/pricing
 * on 2026-09-19. These are standard short-context estimates, not billing receipts.
 * Long-context requests require separate pricing and are rejected below.
 */
export const PRICING: Record<string, { input: number; cachedInput: number; cacheWrite: number; output: number }> = {
  "gpt-5.6-luna": { input: 0.2, cachedInput: 0.02, cacheWrite: 0.25, output: 1.2 },
  "gpt-5.6-sol": { input: 4.0, cachedInput: 0.4, cacheWrite: 5.0, output: 20.0 },
  "text-embedding-3-small": { input: 0.02, cachedInput: 0.02, cacheWrite: 0, output: 0 },
};

/** Ledger token columns. `input_tokens` EXCLUDES cache reads so the four columns are disjoint. */
export interface LedgerUsage {
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  cache_write_tokens: number;
}

function tokenCount(value: number): number {
  if (!Number.isSafeInteger(value) || value < 0) throw new Error("Invalid token usage");
  return value;
}

/** OpenAI's input_tokens includes cached_tokens; split them for the ledger. */
export function toLedgerUsage(usage: OpenAI.Responses.ResponseUsage): LedgerUsage {
  const cached = tokenCount(usage.input_tokens_details?.cached_tokens ?? 0);
  // Newer API responses may supply this field even though the installed SDK predates it.
  const details = usage.input_tokens_details as { cached_tokens?: number; cache_write_tokens?: number };
  const writes = tokenCount(details?.cache_write_tokens ?? 0);
  const input = tokenCount(usage.input_tokens) - cached - writes;
  if (input < 0) throw new Error("Cached tokens exceed total input usage");
  return {
    input_tokens: input,
    output_tokens: tokenCount(usage.output_tokens),
    cache_read_tokens: cached,
    cache_write_tokens: writes,
  };
}

export function estimateCostUsd(model: string, u: LedgerUsage): number {
  if (!Object.hasOwn(PRICING, model)) throw new Error(`No verified pricing for model: ${model}`);
  const p = PRICING[model];
  for (const value of [u.input_tokens, u.output_tokens, u.cache_read_tokens, u.cache_write_tokens]) tokenCount(value);
  if (model === "gpt-5.6-sol" && u.input_tokens + u.cache_read_tokens + u.cache_write_tokens > 272_000) {
    throw new Error("Long-context pricing requires reconciliation");
  }
  return (
    (u.input_tokens * p.input +
      u.cache_read_tokens * p.cachedInput +
      u.cache_write_tokens * p.cacheWrite +
      u.output_tokens * p.output) /
    1_000_000
  );
}

export interface CallModelOptions {
  stage: Stage;
  /** Defaults to MODELS.fast. */
  model?: string;
  /** System prompt. Keep the stable chapter context here, byte-identical across a batch, so cached input hits. */
  instructions?: string;
  /** A string, or a list of {role, content} items. */
  input: string | OpenAI.Responses.ResponseInput;
  /** One key per (document, chapter) so batched calls share a cache prefix. */
  promptCacheKey?: string;
  /** Request strict JSON; callers still validate its schema and semantics. */
  jsonSchema?: { name: string; schema: Record<string, unknown> };
  maxOutputTokens?: number;
  /** Ledger attribution. */
  docId?: string;
  corpusId?: string;
  /** Anything useful for the C5 before/after table: batch size, caching on/off, ... */
  meta?: Record<string, unknown>;
  /** Opt-in process-local result reuse; use off for independent baseline/eval runs. */
  cache?: "off" | "content";
}

export interface CallModelResult {
  text: string;
  response: OpenAI.Responses.Response;
  usage: LedgerUsage;
  costUsd: number;
  latencyMs: number;
  cacheHit?: boolean;
}

let client: OpenAI | undefined;
function openai(): OpenAI {
  // Automatic retries could spend beyond a single request's future reservation.
  return (client ??= new OpenAI({ maxRetries: 0 }));
}

export interface ModelDependencies {
  authorize: () => Promise<void>;
  reserve?: typeof reserveApiSpend;
  settle?: typeof settleApiSpend;
  respond: (request: OpenAI.Responses.ResponseCreateParamsNonStreaming) => Promise<OpenAI.Responses.Response>;
  embeddings: (request: OpenAI.Embeddings.EmbeddingCreateParams) => Promise<OpenAI.Embeddings.CreateEmbeddingResponse>;
  log: (row: LedgerRow) => Promise<void>;
}

const live: ModelDependencies = {
  authorize: async () => {
    await requireLiveBudget();
    if (!process.env.OPENAI_API_KEY) throw new Error("OPENAI_API_KEY is not configured in the process environment");
  },
  reserve: reserveApiSpend,
  settle: settleApiSpend,
  respond: (request) => openai().responses.create(request),
  embeddings: (request) => openai().embeddings.create(request),
  log: (row) => logCall(row),
};

/** Dependency injection supports offline tests; the default transport is spending-gated. */
export function createModelRunner(deps: ModelDependencies) {
const cache = new Map<string, CallModelResult>();
const pending = new Map<string, Promise<CallModelResult>>();
const cachedResult = (result: CallModelResult): CallModelResult => ({ ...structuredClone(result),
  costUsd: 0, latencyMs: 0, cacheHit: true,
  usage: { input_tokens: 0, output_tokens: 0, cache_read_tokens: 0, cache_write_tokens: 0 } });

async function callModel(opts: CallModelOptions): Promise<CallModelResult> {
  if (opts.cache !== "content") return executeModel(opts);
  const key = createHash("sha256").update(JSON.stringify({ model: opts.model ?? MODELS.fast, stage: opts.stage,
    instructions: opts.instructions, input: opts.input, jsonSchema: opts.jsonSchema,
    maxOutputTokens: opts.maxOutputTokens ?? 8192, docId: opts.docId, corpusId: opts.corpusId,
    promptVersion: opts.meta?.prompt_version })).digest("hex");
  const found = cache.get(key);
  if (found) {
    cache.delete(key); cache.set(key, found);
    return cachedResult(found);
  }
  const inFlight = pending.get(key);
  if (inFlight) return cachedResult(await inFlight);
  const work = executeModel(opts);
  pending.set(key, work);
  try {
    const result = await work;
    cache.set(key, structuredClone(result));
    if (cache.size > 128) cache.delete(cache.keys().next().value!);
    return result;
  } finally { pending.delete(key); }
}

async function executeModel(opts: CallModelOptions): Promise<CallModelResult> {
  const model = opts.model ?? MODELS.fast;
  estimateCostUsd(model, { input_tokens: 0, output_tokens: 0, cache_read_tokens: 0, cache_write_tokens: 0 });
  const maxOutput = opts.maxOutputTokens ?? 8192;
  if (!Number.isSafeInteger(maxOutput) || maxOutput <= 0 || maxOutput > 128000) throw new Error("Invalid maxOutputTokens");
  // Restrict the transport to bounded text. No remote media, tools, or conversation IDs.
  if (typeof opts.input !== "string" && !opts.input.every(item =>
    "role" in item && "content" in item && typeof item.content === "string")) throw new Error("Only text model inputs are supported");
  if (Buffer.byteLength(JSON.stringify([opts.input, opts.instructions, opts.jsonSchema])) > 240000) throw new Error("Input exceeds short-context budget");
  await deps.authorize();
  const price = PRICING[model]!;
  // Reserve the entire short-context allowance, including possible cache writes.
  const reservation = await deps.reserve?.((272000 * Math.max(price.input, price.cacheWrite) + maxOutput * price.output) / 1000000, opts.stage, model);
  const started = Date.now();
  const response = await deps.respond({
    model,
    service_tier: "default",
    instructions: opts.instructions,
    input: opts.input,
    max_output_tokens: maxOutput,
    prompt_cache_key: opts.promptCacheKey,
    text: opts.jsonSchema
      ? { format: { type: "json_schema", name: opts.jsonSchema.name, schema: opts.jsonSchema.schema, strict: true } }
      : undefined,
  });
  const latencyMs = Date.now() - started;
  if (!response.usage) throw new Error("Missing usage; billing reconciliation required");
  const usage = toLedgerUsage(response.usage);
  const costUsd = estimateCostUsd(model, usage);

  await deps.log({ stage: opts.stage, model, usage, latencyMs, costUsd, docId: opts.docId, corpusId: opts.corpusId,
    meta: { ...opts.meta, reservation_id: reservation, response_id: response.id, response_status: response.status } });
  if (reservation) await deps.settle!(reservation,
    ((usage.input_tokens + usage.cache_write_tokens) * Math.max(price.input, price.cacheWrite)
      + usage.cache_read_tokens * price.cachedInput + usage.output_tokens * price.output) / 1000000);

  if (response.status !== "completed" || !response.output_text?.trim()) {
    throw new Error("Model response incomplete, refused, or empty");
  }
  if (opts.jsonSchema) {
    try { JSON.parse(response.output_text); } catch { throw new Error("Model returned invalid JSON"); }
  }

  return { text: response.output_text, response, usage, costUsd, latencyMs, cacheHit: false };
}

/** Embeddings for the ES dense index (C2). Logged under stage "embed". */
async function embed(texts: string[], opts: { docId?: string; corpusId?: string; meta?: Record<string, unknown> } = {}): Promise<number[][]> {
  if (texts.length === 0) return [];
  const model = MODELS.embedding;
  estimateCostUsd(model, { input_tokens: 0, output_tokens: 0, cache_read_tokens: 0, cache_write_tokens: 0 });
  await deps.authorize();
  if (texts.some(text => Buffer.byteLength(text) > 8000) || texts.reduce((n, text) => n + Buffer.byteLength(text), 0) > 240000) throw new Error("Embedding input exceeds reservation bounds");
  const reservation = await deps.reserve?.(272000 * PRICING[model]!.input / 1000000, "embed", model);
  const started = Date.now();
  const res = await deps.embeddings({ model, input: texts });
  if (!res.usage) throw new Error("Missing embedding usage; billing reconciliation required");
  const usage: LedgerUsage = {
    input_tokens: tokenCount(res.usage.prompt_tokens),
    output_tokens: 0,
    cache_read_tokens: 0,
    cache_write_tokens: 0,
  };
  await deps.log({
    stage: "embed", model, usage, latencyMs: Date.now() - started, costUsd: estimateCostUsd(model, usage),
    docId: opts.docId, corpusId: opts.corpusId, meta: { ...opts.meta, reservation_id: reservation, count: texts.length },
  });
  if (reservation) await deps.settle!(reservation, estimateCostUsd(model, usage));
  const ordered = [...res.data].sort((a, b) => a.index - b.index);
  const dimensions = ordered[0]?.embedding.length;
  if (!dimensions || ordered.length !== texts.length || ordered.some((d, i) =>
    d.index !== i || d.embedding.length !== dimensions || !d.embedding.every(Number.isFinite))) {
    throw new Error("Invalid embedding response");
  }
  return ordered.map((d) => d.embedding);
}
return { callModel, embed, clearCache: () => cache.clear() };
}

export const { callModel, embed } = createModelRunner(live);

export interface LedgerRow {
  stage: Stage;
  model: string;
  usage: LedgerUsage;
  latencyMs: number;
  costUsd: number;
  docId?: string;
  corpusId?: string;
  meta?: Record<string, unknown>;
}

/** Insert one ledger row. Exported so eval scripts and synthetic tests can log without a real call. */
export async function logCall(row: LedgerRow, query: (sql: string, values: unknown[]) => Promise<unknown> = (sql, values) => db().query(sql, values)): Promise<void> {
  for (const value of [row.usage.input_tokens, row.usage.output_tokens, row.usage.cache_read_tokens, row.usage.cache_write_tokens]) tokenCount(value);
  if (!Number.isFinite(row.costUsd) || row.costUsd < 0 || !Number.isFinite(row.latencyMs) || row.latencyMs < 0) {
    throw new Error("Invalid ledger cost or latency");
  }
  await query(
    `INSERT INTO llm_calls
       (stage, model, input_tokens, output_tokens, cache_read_tokens, cache_write_tokens, latency_ms, cost_usd, corpus_id, doc_id, meta)
     VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)`,
    [
      row.stage,
      row.model,
      row.usage.input_tokens,
      row.usage.output_tokens,
      row.usage.cache_read_tokens,
      row.usage.cache_write_tokens,
      row.latencyMs,
      row.costUsd,
      row.corpusId ?? null,
      row.docId ?? null,
      JSON.stringify(row.meta ?? {}),
    ],
  );
}
